gamestonk_terminal/portfolio/portfolio_model.py

```diff
@@ -333,19 +342,97 @@ def get_rolling_beta(
-        df_var[f"beta_{col}"] = df_var[f"cov_{col}"] / df_var["var"]
-    final = pd.merge(res, df_var, how="left", left_index=True, right_index=True)
-    final = final.fillna(method="ffill")
-    final = final.drop(columns=["var"] + [f"cov_{x}" for x in uniques])
+        exog = sm.add_constant(comb["Close"])
+        rols = RollingOLS(comb[col], exog, window=252)
+        rres = rols.fit()
+        res[f"beta_{col}"] = rres.params["Close"]
+
+    final = res.fillna(method="ffill")
     for uni in uniques:
         final[f"prod_{uni}"] = final[uni] * final[f"beta_{uni}"]
+    dropped = final[[f"beta_{x}" for x in uniques]].copy()
     final = final.drop(columns=[f"beta_{x}" for x in uniques] + uniques)
     final["total"] = final.sum(axis=1)
     final = final[final.index >= datetime.now() - timedelta(days=n + 1)]
-    return final
+    comb = pd.merge(final, dropped, how="left", left_index=True, right_index=True)
+    return comb
+
+
+def get_main_text(df: pd.DataFrame) -> str:
+    """Get main performance summary from a dataframe with returns
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Stock holdings and returns with market returns
+
+    Returns
+    ----------
+    t : str
+        The main summary of performance
+    """
+    v_com = (
+        "greater" if np.var(df["return"]) > np.var(df[("Market", "Return")]) else "less"
+    )
+    d_debt = np.where(df[("Cash", "Cash")] > 0, 0, 1)
+    bcash = 0 if df[("Cash", "Cash")][0] > 0 else abs(df[("Cash", "Cash")][0])
+    ecash = 0 if df[("Cash", "Cash")][-1] > 0 else abs(df[("Cash", "Cash")][-1])
+    bdte = bcash / (df["holdings"][0] - bcash)
+    edte = ecash / (df["holdings"][-1] - ecash)
+    if sum(d_debt) > 0:
+        t_debt = (
+            f"Beginning debt to equity was {bdte:.2%} and ending debt to equity was"
+            f" {edte:.2%}. Debt adds risk to a portfolio by amplifying the gains and losses when"
+            " equities change in value."
+        )
+        if bdte > 1 or edte > 1:
+            t_debt += " Debt to equity ratios above one represent a significant amount of risk."
+    else:
+        t_debt = "Debt was not used this year. This reduces this risk of the portfolio."
+    string = (
+        f"Your portfolio's performance for the period was {df['return'][-1]:.2%}. This was"
+        f" {'greater' if df['return'][-1] > df[('Market', 'Return')][-1] else 'less'} than"
+        f" the market return of {df[('Market', 'Return')][-1]:.2%}. The variance for the"
+        f" portfolio is {np.var(df['return']):.2%}, while the variance for the market was"
+        f" {np.var(df[('Market', 'Return')]):.2%}. Since our portfolio's variance was"
+        f" {v_com} than the market's variance, our returns should be {v_com} than those"
+        f" of the market. {t_debt} The following report details various analytics from the"
+        f" portfolio. Read below to see the moving beta for a stock."
```

Comment:
can you combine this into a single string using` f""" blah blah """`

Reply:
When I tried combining it into a single string like how we do for the terminal print outs it messes up the pdf's spacing. How do I get around this?

Reply:
Not sure.  Haven't played with the pdf library.  Not necessary if it doesn't format well.